.github/workflows/rust.yml

```diff
@@ -3,33 +3,53 @@ on:
   pull_request:
     branches: [develop]
 jobs:
-  build-default:
+  build:
+    strategy:
+      matrix:
+        include:
+          - features: default
+          - features: rusoto-rustls
+            additional_flags: --no-default-features
+          - features: rusoto-native-tls
+            additional_flags: --no-default-features
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v2
+    - uses: actions/cache@v2
+      with:
+        path: |
+          .cargo
+          target
+        # you can edit the .github/cache_bust file if you need to clear the cache
+        key: ${{ hashFiles('.github/cache_bust') }}-${{ runner.os }}-${{ matrix.features }}-${{ hashFiles('**/Cargo.lock') }}
+        restore-keys: |
+          ${{ hashFiles('.github/cache_bust') }}-${{ runner.os }}-${{ matrix.features }}
     - run: rustup update stable
-    - run: cargo build --locked
-    - run: cargo test --locked
-    - run: cargo fmt -- --check
-    - run: cargo clippy --locked -- -D warnings
-  build-rustls:
-    runs-on: ubuntu-latest
-    steps:
-      - uses: actions/checkout@v2
-      - run: rustup update stable
-      - run: cargo build --features rusoto-rustls --no-default-features --locked
-      - run: cargo test --features rusoto-rustls --no-default-features --locked
-  build-native-tls:
-    runs-on: ubuntu-latest
-    steps:
-      - uses: actions/checkout@v2
-      - run: rustup update stable
-      - run: cargo build --features rusoto-native-tls --no-default-features --locked
-      - run: cargo test --features rusoto-native-tls --no-default-features --locked
-  check-licenses:
+    - run: echo "CARGO_HOME=${{ github.workspace }}/.cargo" >> $GITHUB_ENV
+    - run: cargo build --features ${{ matrix.features }} ${{ matrix.additional_flags }} --locked
+    - run: cargo test --features ${{ matrix.features }} ${{ matrix.additional_flags }} --locked
+  check:
```

Comment:
I'm not sure I understand the benefit of separating out the check job; it makes the workflow harder to understand, and we know the fmt/clippy steps will run faster right after a build, and we know the cargo-deny install will be fast with caching; can we run all the steps in one job instead?  fmt/clippy will be repeated, but I think that's correct - since we use features, the code they're checking is actually different each time.  Same with deny, the licenses might be different because of different feature dependencies.